api/agent/pure_runner.go

```diff
@@ -0,0 +1,491 @@
+package agent
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	runner "github.com/fnproject/fn/api/agent/grpc"
+	"github.com/fnproject/fn/api/models"
+	"github.com/go-openapi/strfmt"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+)
+
+// callHandle represents the state of the call as handled by the pure runner, and additionally it implements the
+// interface of http.ResponseWriter so that it can be used for streaming the output back.
+type callHandle struct {
+	engagement runner.RunnerProtocol_EngageServer
+	c          *call // the agent's version of call
+	input      io.WriteCloser
+	started    bool
+	// As the state can be set and checked by both goroutines handling this state, we need a mutex.
+	stateMutex sync.Mutex
+	// Timings, for metrics:
+	receivedTime  strfmt.DateTime // When was the call received?
+	allocatedTime strfmt.DateTime // When did we finish allocating the slot?
+	// Last communication error on the stream (if any). This basically acts as a cancellation flag too.
+	streamError error
+	// For implementing http.ResponseWriter:
+	outHeaders    http.Header
+	outStatus     int
+	headerWritten bool
+}
+
+func (ch *callHandle) Header() http.Header {
+	return ch.outHeaders
+}
+
+func (ch *callHandle) WriteHeader(status int) {
+	ch.outStatus = status
+	ch.commitHeaders()
+}
+
+func (ch *callHandle) commitHeaders() error {
+	if ch.headerWritten {
```

Comment:
this doesn't need the mutex?

Reply:
Only one goroutine reads/write that flag.

Reply:
WriteHeader() is public and this io.Writer is passed out to the actual agent... And proto.Dispatch() runs on it's own go-routine.

Reply:
True... but I'm assuming that only one thread will ever write to the http.ResponseWriter, otherwise the http response would be garbled anyway.

What I mean is, if we get concurrency issues in the writes here, adding the mutex lock would not solve the problem - there would be some broken logic in the program.